src/customfilesystemmodel.cpp

```diff
@@ -169,29 +209,30 @@ bool CustomFileSystemModel::setData(const QModelIndex &index,
         }
         else if(value == Qt::Unchecked)
         {
+            // Get list of files which need manually emitting.  Must be
+            // done before modifying the _checklist.
+            QList<QModelIndex> fakeDescendents = getFakePCRecursive(index);
+
             _partialChecklist.remove(index);
             if(_checklist.remove(index))
-            {
                 emit dataChanged(index, index, selectionChangedRole);
-                if(isDir(index))
-                {
-                    // Set all children to be unchecked.
-                    for(int i = 0; i < rowCount(index); i++)
-                    {
-                        QModelIndex child = index.child(i, index.column());
-                        if(child.isValid())
-                            setData(child, Qt::Unchecked, Qt::CheckStateRole);
-                    }
-                }
-            }
+
+            // Check ancestor
             QModelIndex parent = index.parent();
-            if(parent.isValid())
+            if(parent.isValid() &&
+                    (parent.data(Qt::CheckStateRole) != Qt::Checked))
             {
                 if(hasCheckedSibling(index))
                     setIndexCheckState(parent, Qt::PartiallyChecked);
                 else
                     setIndexCheckState(parent, Qt::Unchecked);
             }
+
+            // Emit data for descendents which falsely reported PC.
+            for (int i = 0; i < fakeDescendents.count(); i++) {
```

Comment:
I merged the single fix in commit https://github.com/Tarsnap/tarsnap-gui/pull/104/commits/48ad531f270f6a739a495891560f3a90f690499d. That's sufficient to fix the internal data discrepancy.

The emits aren't necessary nor can be complete without hitting the recursivity/perf. problem we discussed previously. That's why the viewport update is there, it will take care of updating all items that are in the current view and only that, which is exactly what we need. So the actual FPW will work as expected, even if the emit case for test 13 will still fail. I think we can live with that for the time being.